packages/react-compose/src/createClassResolver.ts

```diff
@@ -14,43 +14,44 @@ import { GenericDictionary, ClassDictionary } from './types';
  * Remaining class names would be interpretted as modifiers, applied to the `root` slot when
  * the component `state` contains a truthy matching prop name.
  */
-export const createClassResolver = (classes: ClassDictionary) => (
-  state: GenericDictionary,
-  slots: GenericDictionary,
-): ClassDictionary => {
-  const classMap: GenericDictionary = {};
-  const modifiers: string[] = [];
+export const createClassResolver = (classes: ClassDictionary, slotNames: string[]) => {
+  // This is in creation time, so this will happen once per css file.
+  const { slots, modifiers, enums } = createResolvedMap(classes, slotNames);
 
-  // Add the default className to root
-  addClassTo(classMap, 'root', state.className);
+  // Everything in the function below will happen at runtime, so it's very critical that this
+  // code is as minimal as possible.
+  // tslint:disable-next-line:no-function-expression
+  return function classResolver(state: GenericDictionary): ClassDictionary {
+    const resolvedClasses: Record<string, string | string[]> = {};
 
-  if (classes) {
-    // Iterate through classes
-    Object.keys(classes).forEach((key: string) => {
-      const classValue = classes[key];
-
-      if (classValue) {
-        // If the class is named the same as a slot, add it to the slot.
-        if (slots.hasOwnProperty(key)) {
-          addClassTo(classMap, key, classValue);
-        } else if (key.indexOf('_') >= 0) {
-          // The class is an enum value. Add if the prop exists and matches.
-          const parts = key.split('_');
-          const enumName = parts[0];
-          const enumValue = parts[1];
+    const modifierClasses = [];
+    for (const modifierName of Object.keys(modifiers)) {
+      if (!!state[modifierName]) {
+        modifierClasses.push(modifiers[modifierName]);
+      }
+    }
 
-          state[enumName] === enumValue && modifiers.push(classValue);
-        } else {
-          state[key] && modifiers.push(classValue);
-        }
+    const enumClasses = [];
+    for (const enumName of Object.keys(enums)) {
+      const enumValueClasses = enums[enumName];
+      if (enumValueClasses[state[enumName]]) {
+        enumClasses.push(enumValueClasses[state[enumName]]);
       }
-    });
+    }
 
-    // Convert the className arrays to strings.
-    Object.keys(classMap).forEach((key: string) => (classMap[key] = classMap[key].concat(modifiers).join(' ')));
-  }
+    for (const slotName of slotNames) {
+      resolvedClasses[slotName] = [
+        slotName === 'root' && state.className ? state.className : '',
+        slots[slotName] || '',
+        ...modifierClasses,
+        ...enumClasses,
```

Comment:
So you're passing all modifier and enum classes to all the slots, correct?

Reply:
yes

Reply:
although we'd get better perf only leaving them at the root